### ISSUE TITLE
[MNG-1775] No property expansion in profile activation

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#3376]
**[Eric Andresen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=eandres)** opened **[MNG-1775](https://issues.apache.org/jira/browse/MNG-1775?redirect=false)** and commented

I have a profile specified in the pom.xml of a project. It is inteded to be activated based on the presence or absence of a file, using the \<file> profile activator.

The profiles are simple:
\<profile>
\<id>metis\</id>
\<activation>
\<file>\<missing>${basedir}/../build.properties\</missing>\</file>
\</activation>
\<build>
\<filters>\<filter>${basedir}/../build.properties.metis\</filter>\</filters>
\</build>
\</profile>
\<profile>
\<id>dev\</id>
\<activation>
\<file>\<exists>${basedir}/../build.properties\</exists>\</file>
\</activation>
\<build>
\<filters>\<filter>${basedir}/../build.properties\</filter>\</filters>
\</build>
\</profile>

The problem comes in with ${basedir} -- it isn't being expanded for purposes of evaluating the file. It's trying to look for a file named "${basedir}/../build.properties", rather than "/home/joe/projectX/projY/../build.properties"; as a result, the "missing" directive is always true, and the dev profile is never activated. When the filter path is evaluated, the ${basedir} property **is** evaluated, however.

---

**Affects:** 2.0, 2.0.1

**Attachments:**
- [basedir_in_activation.patch](https://issues.apache.org/jira/secure/attachment/12712970/basedir_in_activation.patch) (_18.96 kB_)
- [MNG-1775-maven2.2.x.patch](https://issues.apache.org/jira/secure/attachment/12712388/MNG-1775-maven2.2.x.patch) (_18.96 kB_)
- [test.zip](https://issues.apache.org/jira/secure/attachment/12712546/test.zip) (_2.70 kB_)
- [test2.zip](https://issues.apache.org/jira/secure/attachment/12712547/test2.zip) (_4.82 kB_)

**Issue Links:**
- [MNG-2363](https://issues.apache.org/jira/browse/MNG-2363) \<profile>\<activation>\<file>\<exists/> does not work in a multi-project build
 (_**"duplicates"**_)
- [MNG-2809](https://issues.apache.org/jira/browse/MNG-2809) Can't activate a profile by checking for the presence of a file in ${user.home}
 (_**"duplicates"**_)
- [MNG-2627](https://issues.apache.org/jira/browse/MNG-2627) properties substitutions in exists
 (_**"is duplicated by"**_)

10 votes, 9 watchers

**Comment by jira-importer** [comment]
**[Eric Andresen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=eandres)** commented

I originally did not use ${basedir} at all, but that does not function when the build is executed as a subproject. The origination of the file presence activation, as defined by MNG-821, recommended the use of ${basedir} to get around this, which of course, isn't working.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

I'm also seeing this in settings.xml:

\<server>
\<username>${username}\</username>
...
\</server>
\<profile>
...
\<properties>
\<username>notbrett\</username>
\</properties>
\<id>property-overrides\</id>
\</profile>
\<activeProfiles>
\<activeProfile>property-overrides\</activeProfile>
\</activeProfiles>

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

ok, I can't reproduce my issue since fixing MNG-1509

I can reproduce the original issue, but it would require futzing with ordering in the project builder I expect, so I'm not going to rush that into 2.0.3.

**Comment by jira-importer** [comment]
**[Mike Youngstrom](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=youngm)** commented

I am experiencing the same problem but for system properties too.

Below does not work.
mvn package -Dbuild.env=prod
\<file>
\<exists>src/${build.env}/webapp/WEB-INF/web.xml\</exists>
\</file>

If I activate based on the existence of build.env being prod that does work.

This works also.
\<file>
\<exists>src/prod/webapp/WEB-INF/web.xml\</exists>
\</file>

It appears it's just not parsing system properties.

Mike

**Comment by jira-importer** [comment]
**[Mike Youngstrom](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=youngm)** commented

One other note.  I don't know if it matters but the above example is in a parent pom.

**Comment by jira-importer** [comment]
**[James Olsen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jim_b_o)** commented

I think I have a similar problem.  I have 'version' properties defined in settings.xml which aren't always expanded, e.g:

<<<<<
[INFO] -------------------------------------------------------------
[ERROR] BUILD ERROR
[INFO] -------------------------------------------------------------
[INFO] Failed to resolve artifact.

GroupId: com.cognotec.common
ArtifactId: cognotec-common
Version: ${cognotec-common.version}

Reason: Unable to download the artifact from any repository

com.cognotec.common:cognotec-common:pom:${cognotec-common.version}

> ... where cognotec-common.version has been defined in the settings.xml in an active profile.
> 
> If I build from a top level aggregator pom, it works fine.  If I build an individual module the expansion fails as shown above.
> 
> I use a flat directory structure with 'logical' parent directories:
> 
> trunk/build/pom.xml (master pom for inheritance)
> 
> trunk/subsystem1/subsystem1-root-module/pom.xml (inherits from ../../build/pom.xml and aggregates module1 and module2)
> trunk/subsystem1/module1/pom.xml (inherits from ../subsystem1-root-module/pom.xml)
> trunk/subsystem1/module2/pom.xml (inherits from ../subsystem1-root-module/pom.xml)
> 
> trunk/subsystem2/subsystem2-root-module/pom.xml (inherits from ../../build/pom.xml and aggregates module3 and module4)
> trunk/subsystem2/module3/pom.xml (inherits from ../subsystem2-root-module/pom.xml)
> trunk/subsystem2/module4/pom.xml (inherits from ../subsystem2-root-module/pom.xml)
> 
> trunk/pom.xml (aggregates build, subsystem1-root-module and subsystem2-root-module)
> 
> Works if I build trunk/pom.xml or one of the subsystemX-root-module/pom.xml files but not if I build a leaf module directly.

**Comment by jira-importer** [comment]
**[James Olsen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jim_b_o)** commented

Forgot to mention that in every case the result of 'mvn help:effective-pom' always returns a result with all the properties successfully expanded.

**Comment by jira-importer** [comment]
**[Shelley](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shelley)** commented

I am encountering this issue as well using profiles in Maven 2.0.9 on Windows Vista with an individual JAR project (it is not a parent/reactor build, nor does it have a parent).  I need to activate a profile based on the existence of a packaged jar in the build output directory (ie: ${project.build.directory}/${project.build.finalName}).  Hard-coding the build directory, though undesirable, is not a problem, but hard-coding the final name is error-prone since this will need to manually change during/after every release.

It looks like there hasn't been any activity on this bug for several years; are there plans to resolve this, or is this limitation intentional?

**Comment by jira-importer** [comment]
**[Zachary Smith](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zaque)** commented

I'd also like to chime in and request a fix to this.  I've been pulling my hair out on this all morning.   Hard coding the path does work but will not be easily used in our environment so I'm sort of stuck.

The profile is supposed to run to make sure that no dev domain is running when the CI tests run because CI generates its own domain.

    <profile>
         
         <id>ci-setup</id>
         <activation>
              <file>
                                  
                   <exists>${dev.domain.dir}</exists>
              </file>
         </activation>
         <build>
              <plugins>
                   <plugin>
                            <groupId>org.glassfish.maven.plugin</groupId>
                            <artifactId>maven-glassfish-plugin</artifactId>
                            <executions>
                                     <execution>
                                              <id>stop-domain</id>
                                              <inherited>true</inherited>
                                              <goals>
                                                       <goal>stop-domain</goal>
                                              </goals>
                                              <phase>pre-integration-test</phase>
                                     </execution>
                            </executions>
                   </plugin>
              </plugins>
         </build>
    </profile>

**Comment by jira-importer** [comment]
**[Karel Vervaeke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karel1980)** commented

Attached a simple test case.

mvn install -> build fails
mkdir '${basedir}'; touch '${basedir}/pom.xml'; mvn install -> build succeeds

**Comment by jira-importer** [comment]
**[Karel Vervaeke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karel1980)** commented

Does anyone have a good workaround for this?
How do you activate a profile based on the existance of a file whose path is relative to the pom?

**Comment by jira-importer** [comment]
**[Marc Rohlfs](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=marc.rohlfs)** commented

Hi Karel, did You try it without using the `basedir` property?

```xml
<activation>
  <activeByDefault>false</activeByDefault>
  <file>
    <missing>pom.xml</missing>
  </file>
</activation>
```

Just verified this with Maven 2.0.10 and 2.2.1.

I remember having seen an expansion for (only) the `basedir` property being implemented for file activation, but I don't remember where exactly it was - and I don't have a checkout right here. Guess it was somewhere in Maven 3 code.

**Comment by jira-importer** [comment]
**[Karel Vervaeke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karel1980)** commented

Without the basedir property it works, however, the test would not work when building a multi-module project:

myproject
test (module)

When running from 'myproject', the file myproject/pom.xml would be checked
When running from 'test', the file myproject/test/pom.xml would be checked.

(Of course the example could be better since the pom.xml file would always be present.)

**Comment by jira-importer** [comment]
**[Karel Vervaeke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karel1980)** commented

I just spent some quality time with my debugger.  Here are some of my conclusions (in the order of conception)

* it appears that the interpolation of ${whatever} is performed,
  but testing the profile activation happens **BEFORE** the interpolation.

One possible solution would go along these lines:
1. originalModel = cloneModel(model)
2. interpolate(model)
3. determine active profiles, add their properties to the list of properties for interpolation
4. interpolate(model) again.

But then you would have weird situations where a profile tests for ${project.build.directory}/something and at the same time
it modifies ${project.build.directory} (by declaring \<build>\<directory>...\</directory>\</build>

* looking somewhat further into FileProfileActivator, I found that some interpolation is
  done on the \<exists> and \<missing> values but afaict only these are used for interpolation of environment variables and system properties

* FileProfileActivator is using RegexpBasedInterpolator.  My first idea would be to
  change the Activator interface so it takes a java.util.Properties with additional properties to doe interpolation on.
  "basedir" would be a property that's easy enough to add to this map.

That's it for now, I'm off until wednesday next week - Will give a shot at providing a patch then.

**Comment by jira-importer** [comment]
**[Karel Vervaeke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karel1980)** commented

Attached improved test project (which fails when ${basedir} is not expanded, and succeeds when ${basedir} is expanded).

Attached patch which passes a properties object containing a key for 'basedir' to the activators.
The file activator uses the properties for expanding \<exists> and \<missing> files.

**Comment by jira-importer** [comment]
**[Marc Rohlfs](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=marc.rohlfs)** commented

Hi Karel, just some comments:

> Without the basedir property it works, however, the test would not work when building a multi-module project:
> 
> myproject
> test (module)
> 
> When running from 'myproject', the file myproject/pom.xml would be checked
> When running from 'test', the file myproject/test/pom.xml would be checked.

The `basedir` property always points to the directory of the POM that is currently processed, not depending if You initiate the build from a parent module. If You want to check the `myproject/pom.xml` in `myproject/test/pom.xml`, You should use another property (like e.g. `multiproject.basedir`) and define it like

```xml
<multimodule.basedir>${basedir}<multimodule.basedir>
```

in the parent POM and

```xml
<multimodule.basedir>${basedir}/..<multimodule.basedir>
```

in the 1st level submodules and so on ...
This means that it wouldn't help You to work on the expansion of just the `basedir` property. I didn't check the patch You attached today, but I wonder if it worked.

> One possible solution would go along these lines:
> 1. originalModel = cloneModel(model)
> 2. interpolate(model)
> 3. determine active profiles, add their properties to the list of properties for interpolation
> 4. interpolate(model) again.
> 
> But then you would have weird situations ...

I'm totally with You on this, I'd really like to have a two-step interpolation just like You described it. But I also see undefined situation coming up. And I think that's the reason why (by now) they only consider the expansion of the `basedir` property for profile activations. I'd suggest to discuss this on the Maven Developers mailing list first, because otherwise Maven Comitters might just decline any patch You invent on this.

About the patch You attached: Was it really a patch for the trunk (respectively for Maven 3)? Somehow I think that it is for Maven 2 (`maven-2/branches/maven-2.2.x`). Did You read the page about [Creating and submitting a patch](http://maven.apache.org/guides/development/guide-m2-development.html#Creating_and_submitting_a_patch)?

Kind regards
Marc

**Comment by jira-importer** [comment]
**[Karel Vervaeke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karel1980)** commented

The patch I submitted was for maven-2.2.x - Sorry, should have been clearer on that.  I'll go ahead and jump on the maven developers list.
for completeness: the patch works as you described, expanding ${basedir} to the directory of the pom that contains the ${basedir} property.

**Comment by jira-importer** [comment]
**[Karel Vervaeke](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=karel1980)** commented

Renamed the patch to be more compliant with the suggested patch naming scheme.

**Comment by jira-importer** [comment]
**[Reiner Saddey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rsaddey)** commented

Fix has been postponed to 3.x :-(

Might it be appropriate to apply the fix to 2.2.x as well, ${basedir} being  explicitly stated within current 2.x docs at http://maven.apache.org/pom.html#Activation ?
